File: skincare-routines/src/chat/local/store.ts
/** Browser twin of chat_api/data/store.py: the single read model the tools use. It reads the very same public/data
 * files the pages render from (through the shared fetch-once cache), so the assistant can never disagree with the UI. */
import { loadJson } from '../../data/fetchJson';
import { buildIndex, type CategoryIndex } from '../../domain/index';
import type { Benchmark, CategoryData, CategoryMeta, InciColumns, IngredientAlias, KnowledgeData, Manifest, ProductDetail, ProductRow, RoutinesData } from '../../lib/types';
import { SearchIndex } from './search';
import { loadInciColumns, loadSearchColumns } from './searchFile';

export class DataError extends Error {}

/** A loaded category with the lookups the tools need on top of the UI's own index. */
export class CategoryView {
  readonly id: string;
  readonly items: ProductRow[];
  readonly tagIndex: string[];
  readonly facets: CategoryIndex['facets'];
  readonly ranks: Uint32Array;
  readonly tagPos: Map<string, number>;
  readonly posOf = new Map<string, number>();
  readonly byRank: number[];

  constructor(idx: CategoryIndex) {
    this.id = idx.id; this.items = idx.items; this.tagIndex = idx.tagIndex; this.facets = idx.facets;
    this.ranks = idx.rank; this.tagPos = idx.tagPos;
    idx.items.forEach((it, i) => this.posOf.set(it.id, i));
    this.byRank = idx.items.map((_, i) => i).sort((a, b) => idx.rank[a] - idx.rank[b]);
  }

  tagsOf(item: ProductRow) { return item.t.map((t) => this.tagIndex[t]); }
  hasTags(item: ProductRow, wanted: number[]) { return wanted.every((t) => item.t.includes(t)); }
  /** Site filter-panel semantics: each group is one constraint (any of its tags, or all of them for `and` groups). */
  matchesTagGroups(item: ProductRow, groups: TagGroupQuery[]) {
    return groups.every((g) => (g.all ? g.tags.every((t) => item.t.includes(t)) : g.tags.some((t) => item.t.includes(t))));
  }
}

export interface TagGroupQuery { group: string; tags: number[]; all: boolean }

function shardOf(id: string, shards: number) {
  let h = 0;
  for (let i = 0; i < id.length; i++) h = (h * 31 + id.charCodeAt(i)) >>> 0;
  return h % shards;
}

export class LedgerStore {
  private readonly base: string;
  private readonly categories = new Map<string, CategoryView>();
  private readonly inciCols = new Map<string, Promise<InciColumns>>();
  private searchPromise: Promise<SearchIndex> | null = null;
  private manifestCache: Manifest | null = null;

  constructor(base = `${import.meta.env.BASE_URL}data/`) { this.base = base; }

  async manifest(): Promise<Manifest> {
    this.manifestCache = await loadJson<Manifest>(`${this.base}manifest.json`);
    return this.manifestCache;
  }

  categoryMeta(id: string): CategoryMeta | null {
    return this.manifestCache?.categories.find((c) => c.id === id) ?? null;
  }

  benchmarkFor(id: string): Benchmark | null {
    return this.manifestCache?.benchmarks.find((b) => b.category === id) ?? null;
  }

  async category(id: string): Promise<CategoryView> {
    await this.manifest();
    const cached = this.categories.get(id);
    if (cached) return cached;
    if (!this.categoryMeta(id)) throw new DataError(`No ranked category called '${id}'`);
    const view = new CategoryView(buildIndex(await loadJson<CategoryData>(`${this.base}${id}.json`)));
    this.categories.set(id, view);
    return view;
  }

  async detail(categoryId: string, productId: string): Promise<ProductDetail | null> {
    const m = await this.manifest();
    const shard = await loadJson<Record<string, ProductDetail>>(`${this.base}${categoryId}.d${shardOf(productId, m.shards)}.json`);
    return shard[productId] ?? null;
  }

  routines(): Promise<RoutinesData> { return loadJson<RoutinesData>(`${this.base}routines.json`); }

  /** The scorer's sourced ingredient tables + pairing guidance; absent on datasets generated before knowledge.json existed. */
  async knowledge(): Promise<KnowledgeData> {
    const m = await this.manifest();
    if (!m.knowledge) throw new DataError('This dataset was generated without the ingredient knowledge file (knowledge.json).');
    return loadJson<KnowledgeData>(`${this.base}${m.knowledge.file}`);
  }

  /** Normalised INCI / seller-ingredient columns for one category, aligned with `CategoryView.items` (fetched once). */
  inci(categoryId: string): Promise<InciColumns> {
    const cached = this.inciCols.get(categoryId);
    if (cached) return cached;
    const p = this.manifest().then(async (m) => {
      if (!m.inci) throw new DataError('This dataset was generated without ingredient columns (<category>.inci.json) — rebuild the data.');
      const cols = await loadInciColumns(`${this.base}${categoryId}.${m.inci.suffix}`, `${this.base}${categoryId}.${m.inci.gzip}`);
      if (cols.generatedAt !== m.generatedAt) throw new DataError(`Ingredient columns for ${categoryId} (${cols.generatedAt}) and manifest (${m.generatedAt}) are from different builds — reload the page.`);
      return cols;
    });
    this.inciCols.set(categoryId, p);
    p.catch(() => { this.inciCols.delete(categoryId); });
    return p;
  }

  /** Common-name → INCI alias table from knowledge.json (empty on older datasets, so plain terms still work). */
  async ingredientAliases(): Promise<IngredientAlias[]> {
    const m = await this.manifest();
    if (!m.knowledge) return [];
    return (await loadJson<KnowledgeData>(`${this.base}${m.knowledge.file}`)).ingredientAliases ?? [];
  }

  /** The cross-category index is a few MB, so it is fetched once, on the first tool that needs it. */
  search(): Promise<SearchIndex> {
    if (!this.searchPromise) {
      this.searchPromise = this.manifest().then(async (m) => {
        if (!m.search) throw new DataError('This dataset was generated without a search index (search.json).');
        const cols = await loadSearchColumns(`${this.base}${m.search.file}`, `${this.base}${m.search.gzip}`);
        if (cols.generatedAt !== m.generatedAt) throw new DataError(`Search index (${cols.generatedAt}) and manifest (${m.generatedAt}) are from different builds — reload the page.`);
        return new SearchIndex(cols);
      });
      this.searchPromise.catch(() => { this.searchPromise = null; });
    }
    return this.searchPromise;
  }
}

File: skincare-routines/chat-api/chat_api/data/store.py
"""LedgerStore: the single read model over the generated dataset.

* Versioned by `manifest.generatedAt` — when the site regenerates data, the next refresh drops every cache and
  rebuilds the search index, so new categories/listings are picked up without a redeploy (Observer: listeners
  are notified of the new version).
* Category compact files and detail shards are loaded lazily into small LRU caches (Proxy / lazy loading).
"""
from __future__ import annotations

import asyncio
import time
from collections import OrderedDict
from collections.abc import Callable
from typing import Any

from .search import SearchIndex
from .source import DataError, JsonSource


class LRU:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self._items: OrderedDict[str, Any] = OrderedDict()

    def get(self, key: str) -> Any | None:
        if key not in self._items:
            return None
        self._items.move_to_end(key)
        return self._items[key]

    def put(self, key: str, value: Any) -> None:
        self._items[key] = value
        self._items.move_to_end(key)
        while len(self._items) > self.capacity:
            self._items.popitem(last=False)

    def clear(self) -> None:
        self._items.clear()

    def __len__(self) -> int:
        return len(self._items)


def rank_order(items: list[dict]) -> list[int]:
    """1-based overall rank per position: score desc, then price asc — identical to the frontend's `buildIndex`."""
    order = sorted(range(len(items)), key=lambda i: (-items[i]["s"], items[i]["p"]))
    ranks = [0] * len(items)
    for r, pos in enumerate(order):
        ranks[pos] = r + 1
    return ranks


def shard_of(product_id: str, shards: int) -> int:
    h = 0
    for ch in product_id:
        h = (h * 31 + ord(ch)) & 0xFFFFFFFF
    return h % shards


class CategoryView:
    """A loaded compact category with its rank array and tag lookup."""

    def __init__(self, data: dict):
        self.id: str = data["id"]
        self.items: list[dict] = data["items"]
        self.tag_index: list[str] = data["tagIndex"]
        self.facets: dict[str, list[dict]] = data["facets"]
        self.ranks = rank_order(self.items)
        self.tag_pos = {tag: i for i, tag in enumerate(self.tag_index)}
        self.pos_of = {item["id"]: i for i, item in enumerate(self.items)}
        self.by_rank = sorted(range(len(self.items)), key=lambda i: self.ranks[i])

    def tags_of(self, item: dict) -> list[str]:
        return [self.tag_index[t] for t in item["t"]]

    def has_tags(self, item: dict, wanted: list[int]) -> bool:
        have = set(item["t"])
        return all(t in have for t in wanted)


class LedgerStore:
    def __init__(self, source: JsonSource, *, refresh_seconds: int, category_cache: int, shard_cache: int):
        self.source = source
        self.refresh_seconds = refresh_seconds
        self.manifest: dict | None = None
        self.version: str | None = None
        self.routines: dict | None = None
        self.knowledge: dict | None = None
        self.index = SearchIndex()
        self._categories = LRU(category_cache)
        self._shards = LRU(shard_cache)
        self._checked_at = 0.0
        self._lock = asyncio.Lock()
        self._listeners: list[Callable[[str], None]] = []
        self.last_error: str | None = None
        self.indexed_at: float | None = None

    # ---- versioning -------------------------------------------------------------------------------------------
    def on_version(self, listener: Callable[[str], None]) -> None:
        self._listeners.append(listener)

    async def ensure_fresh(self, force: bool = False) -> dict:
        """Re-read the manifest at most every `refresh_seconds`; rebuild everything when `generatedAt` changed."""
        now = time.monotonic()
        if self.manifest is not None and not force and now - self._checked_at < self.refresh_seconds:
            return self.manifest
        async with self._lock:
            if self.manifest is not None and not force and time.monotonic() - self._checked_at < self.refresh_seconds:
                return self.manifest
            try:
                manifest = await self.source.read("manifest.json")
            except DataError as exc:
                self.last_error = str(exc)
                self._checked_at = time.monotonic()
                if self.manifest is None:
                    raise
                return self.manifest
            self._checked_at = time.monotonic()
            version = manifest.get("generatedAt")
            if version != self.version:
                await self._rebuild(manifest)
            return self.manifest or manifest

    async def _rebuild(self, manifest: dict) -> None:
        index = SearchIndex()
        for cat in manifest.get("categories", []):
            data = await self.source.read(f"{cat['id']}.json")
            index.add_category(cat["id"], data["items"], rank_order(data["items"]))
        self.manifest = manifest
        self.version = manifest.get("generatedAt")
        self.index = index
        self.routines = None
        self.knowledge = None
        self._categories.clear()
        self._shards.clear()
        self.indexed_at = time.time()
        self.last_error = None
        for listener in self._listeners:
            listener(self.version or "")

    # ---- reads ------------------------------------------------------------------------------------------------
    def category_meta(self, category_id: str) -> dict | None:
        if not self.manifest:
            return None
        return next((c for c in self.manifest["categories"] if c["id"] == category_id), None)

    def benchmark_for(self, category_id: str) -> dict | None:
        if not self.manifest:
            return None
        return next((b for b in self.manifest.get("benchmarks", []) if b["category"] == category_id), None)

    async def category(self, category_id: str) -> CategoryView:
        await self.ensure_fresh()
        cached = self._categories.get(category_id)
        if cached is not None:
            return cached
        if not self.category_meta(category_id):
            raise DataError(f"No ranked category called '{category_id}'")
        view = CategoryView(await self.source.read(f"{category_id}.json"))
        self._categories.put(category_id, view)
        return view

    async def detail(self, category_id: str, product_id: str) -> dict | None:
        manifest = await self.ensure_fresh()
        shards = int(manifest.get("shards", 1))
        key = f"{category_id}.d{shard_of(product_id, shards)}.json"
        shard = self._shards.get(key)
        if shard is None:
            shard = await self.source.read(key)
            self._shards.put(key, shard)
        return shard.get(product_id)

    async def get_routines(self) -> dict:
        await self.ensure_fresh()
        if self.routines is None:
            self.routines = await self.source.read("routines.json")
        return self.routines

    async def get_knowledge(self) -> dict:
        """The scorer's sourced ingredient tables + pairing guidance; absent on datasets generated before knowledge.json."""
        manifest = await self.ensure_fresh()
        if self.knowledge is None:
            meta = manifest.get("knowledge")
            if not meta:
                raise DataError("This dataset was generated without the ingredient knowledge file (knowledge.json).")
            self.knowledge = await self.source.read(meta["file"])
        return self.knowledge

    def stats(self) -> dict:
        return {
            "version": self.version,
            "indexed": len(self.index),
            "categoriesLoaded": len(self._categories),
            "shardsLoaded": len(self._shards),
            "source": self.source.label,
            "lastError": self.last_error,
            "indexedAt": self.indexed_at,
        }
